ssh/src/extension.ts

```diff
@@ -46,41 +53,226 @@ export function activate(context: vscode.ExtensionContext): void {
-	conn?.destroy();
+	destination?.destroy();
 }
 
 function createForwardedDisplay(conn: Client, options: ConnectOptions): Promise<string> {
 	return new Promise((resolve, reject) => {
-		const { username } = options;
-		const host = getServerHost() || options.host;
-		const port = getServerPort() || options.port;
+		logger.log("started");
+
+		conn
+			.on('x11', handleX11)
+			.on('ready', () => {
+				// Create an interactive shell with X11 forwarding and leave it open
+				// as long as VS Code is running so that we can point VS Code to its
+				// display.
+				logger.log("Creating forwarding shell");
+				resolve(createForwardingShell(conn, true));
+			})
+			.on('error', (err) => {
+				reject(err);
+			})
+			.on('close', () => {
+				if (jump && jumpOptions) {
+					logger.log(`Closing ${getHostString(jumpOptions)} along with ${getHostString(destinationOptions)}`);
+					jump.end();
+				}
+			});
+		
+		jump
+			.on('ready', () => {
+				// Create an interactive shell with X11 forwarding and leave it open
+				// as long as VS Code is running so that we can point VS Code to its
+				// display.
+				createForwardingShell(jump, false);
+
+				jump.forwardOut('127.0.0.1', 12345, destinationOptions.host, destinationOptions.port, (err, stream) => {
+					if (err) {
+						reject(err);
+						return jump.end();
+					}
+					
+					logger.log(`Connecting to ${getHostString(destinationOptions)}`);
+					destination.connect({
+						sock: stream,
+						...getConnectConfig(destinationOptions)
+					});
+				});
+			})
+			.on('error', (err) => {
+				reject(err);
+			});
+
+		const preferConfig = getPreferConfig();
+		const jumpHost = getJumpHost();
+		let destinationOptions: ConnectOptions  = { 
+			username: options.username,
+			host: getServerHost() || options.host,
+			port: getServerPort() || options.port
+		}
 
-		logger.log(`Connecting to ${username}@${host} port ${port}`);
+		// Populate jump host falling back on server settings for user and port
+		let jumpOptions: ConnectOptions | null = null;
+		if (jumpHost) {
+			const jumpUser = getJumpUser();
+			const jumpPort = getJumpPort();
+			jumpOptions = {
+				username: jumpUser ? jumpUser : options.username,
+				host: jumpHost,
+				port: jumpPort ? jumpPort: options.port
+			}
+		}
 
-		conn.on('x11', handleX11);
+		if (preferConfig) {
+			const sshConfig = resolveHome(getSshConfig());
+			const host = getServerHost() || options.host;
+
+			logger.log(`Attempting to parse ${sshConfig} for ssh configuration of ${host}`);
+
+			try {
+				const data = fs.readFileSync(resolveHome(getSshConfig()))
+
+				const config = SSHConfig.parse(data.toString());
+
+				let hostConfig = null;
+				for (const line of config) {
+					if (line.param === 'Host') {
+						if (line.value === host) {
+							hostConfig = line.config;
+							break;
+						} else {
+							const hostNameMatch = line.config.find((line:any) => line.param === 'HostName' && line.value === host);
```

Comment:
Is the `any` because there aren't any TS types for ssh-config yet? Ideally we'd either write some and contribute them to ssh-config itself if the author wants them or https://github.com/DefinitelyTyped/DefinitelyTyped if not. 

A quicker way to improve this though would be to create an `ssh-config.d.ts` file in this repo and define just the types we need in it. It would probably look something like this:

```ts
declare module 'ssh-config' {
    interface ConfigSection { 
        // ...
    }

    interface ConfigLine {
        param: string;
        value: string;
    }

    interface FindOptions {
        Host?: string;
        Match?: string;
    }

    class SSHConfig {
        static parse(config: string): SSHConfig;

        find(opts: FindOptions | ((line: ConfigLine) => boolean)): ConfigSection;
        // other functions...
    }

    export = SSHConfig;
}
```